WordPress/Classes/Services/Store.swift

```diff
@@ -22,53 +22,147 @@ class StoreKitTransactionObserver: NSObject, SKPaymentTransactionObserver {
+            return
+        }
+        
+        pendingPayment = (plan.id, product.productIdentifier, siteID)
         store.requestPayment(product)
     }
 
     func processTransaction(transaction: SKPaymentTransaction) {
         DDLogSwift.logInfo("[Store] Processing transaction \(transaction)")
-        guard transaction.transactionState == .Purchased else {
-            // Ignore other states for now
-            if transaction.transactionState == .Failed {
-                DDLogSwift.logInfo("[Store] Finishing failed transaction \(transaction)")
-                SKPaymentQueue.defaultQueue().finishTransaction(transaction)
-            } else {
-                DDLogSwift.logInfo("[Store] Ignoring transaction \(transaction)")
-            }
+        switch transaction.transactionState {
+        case .Purchasing: break
+        case .Restored: break
+        case .Failed:
+            DDLogSwift.logInfo("[Store] Finishing failed transaction \(transaction)")
+            finishTransaction(transaction)
+        case .Deferred:
+            DDLogSwift.logInfo("[Store] Transaction is deferred \(transaction)")
+        case .Purchased:
+            verifyTransaction(transaction)
+        }
+    }
+    
+    private func verifyTransaction(transaction: SKPaymentTransaction) {
+        guard let pendingPayment = pendingPayment else {
+            DDLogSwift.logInfo("[Store] Transaction with no pending payment information \(transaction)")
+            
+            // TODO: (@frosty 2016-04-27) Still attempt to verify purchase, sending only user info /
+            // receipt data – we should at least be able to tell if this is a renewal.
+            
+            finishTransaction(transaction)
             return
         }
-        assert(transaction.payment.productIdentifier == pendingPayment?.0.productIdentifier)
-        guard let siteID = pendingPayment?.siteID,
-            let plan = pendingPayment?.plan,
-            let service = PlanService(siteID: siteID, store: StoreKitStore()),
+        
+        assert(transaction.payment.productIdentifier == pendingPayment.productID)
+        
+        guard let service = PlanService(siteID: pendingPayment.siteID, store: StoreKitStore()),
             let receiptURL = NSBundle.mainBundle().appStoreReceiptURL,
             let receipt = NSData(contentsOfURL: receiptURL)
             else {
                 assertionFailure()
                 return
         }
+        
         DDLogSwift.logInfo("[Store] Verifying purchase for transaction \(transaction)")
-        service.verifyPurchase(siteID, plan: plan, receipt: receipt, completion: { _ in
-            DDLogSwift.logInfo("[Store] Finishing transaction \(transaction)")
-            SKPaymentQueue.defaultQueue().finishTransaction(transaction)
+        service.verifyPurchase(pendingPayment.siteID, planID: pendingPayment.planID, receipt: receipt, completion: { [weak self] _ in
+            self?.finishTransaction(transaction)
```

Comment:
This probably belongs in the REST implementation, but just leaving a note here: if this request fails we should first retry, and then not finish the transaction until we know the server has processed it.